File: takeshi_bot/commands/member/suporte.py
from __future__ import annotations

import base64
from pathlib import Path
from typing import Any

try:
    from openai import AsyncOpenAI
except ModuleNotFoundError:  # pragma: no cover - exercised by handle when needed.
    AsyncOpenAI = None  # type: ignore[assignment]

from takeshi_bot import config
from takeshi_bot.commands import Command
from takeshi_bot.context import CommandContext
from takeshi_bot.errors import DangerError, WarningError
from takeshi_bot.paths import PROJECT_ROOT
from takeshi_bot.utils import random_name, remove_file_if_exists


def _context_text() -> str:
    parts: list[str] = []
    for name in ("AGENTS.md", "README.md", "package.json"):
        path = PROJECT_ROOT / name
        if path.exists():
            parts.append(path.read_text(encoding="utf-8", errors="replace")[:12000])
    return "\n\n".join(parts)


def _support_text(ctx: CommandContext) -> str:
    has_args = bool(ctx.args)
    text = ctx.full_args.strip() if has_args else ctx.reply_text.strip()
    if has_args and ctx.reply_text:
        return f"Contexto anterior: {ctx.reply_text}\n\nNova questao: {text}"
    return text


def _image_mime_type(image_path: str) -> str:
    extension = Path(image_path).suffix.lower()
    return {
        ".png": "image/png",
        ".jpg": "image/jpeg",
        ".jpeg": "image/jpeg",
        ".webp": "image/webp",
        ".gif": "image/gif",
    }.get(extension, "image/jpeg")


def _image_content_part(image_path: str) -> dict[str, Any]:
    encoded = base64.b64encode(Path(image_path).read_bytes()).decode("ascii")
    return {
        "type": "image_url",
        "image_url": {
            "url": f"data:{_image_mime_type(image_path)};base64,{encoded}",
            "detail": "low",
        },
    }


async def handle(ctx: CommandContext) -> None:
    if not config.OPENAI_API_KEY:
        raise WarningError("O suporte inteligente nao esta disponivel no momento.")
    if ctx.is_video:
        raise WarningError("Nao consigo interpretar videos ainda! Envie imagem ou texto.")
    if ctx.is_audio:
        raise WarningError("Nao consigo interpretar audios ainda! Envie imagem ou texto.")

    text = _support_text(ctx)
    if not text and not ctx.is_image:
        await ctx.send_react(config.BOT_EMOJI)
        await ctx.send_reply(
            "*Takeshi Suporte*\n\n"
            "Faca sua pergunta sobre mim que eu te ajudarei!\n\n"
            "- /suporte bot desliga sozinho\n"
            "- /suporte como instalar no Termux?\n"
            "- /suporte erro 401 API Spider X\n"
            "- Envie uma imagem com /suporte para analise visual"
        )
        return
    if text and len(text) < 5:
        raise DangerError("O texto deve ter no minimo 5 caracteres.")
    if text and len(text) > 4096:
        raise DangerError("O texto deve ter no maximo 4096 caracteres.")

    await ctx.send_wait_reply("Analisando sua pergunta...")
    image_path = await ctx.download_image(random_name("support")) if ctx.is_image else None
    try:
        if AsyncOpenAI is None:
            raise WarningError("A dependencia openai nao esta instalada no ambiente.")
        client = AsyncOpenAI(api_key=config.OPENAI_API_KEY)
        user_content: list[dict[str, Any]] = []
        if text:
            user_content.append({"type": "text", "text": text})
        if image_path:
            if not text:
                user_content.append({"type": "text", "text": "O que voce ve nesta imagem?"})
            user_content.append(_image_content_part(image_path))
        if not user_content:
            user_content.append({"type": "text", "text": "O que voce ve nesta imagem?"})
        messages: list[dict[str, Any]] = [
            {
                "role": "system",
                "content": (
                    "Voce e um assistente especializado em suporte tecnico do Takeshi Bot. "
                    "Responda em portugues do Brasil, direto e com foco pratico."
                ),
            },
            {"role": "system", "content": _context_text()},
            {"role": "user", "content": user_content},
        ]
        response = await client.chat.completions.create(
            model="gpt-5-mini",
            messages=messages,
            max_completion_tokens=2048,
        )
        answer = (response.choices[0].message.content or "").strip()
        if not answer:
            raise DangerError("Nao consegui encontrar uma resposta para sua pergunta.")
        await ctx.send_react(config.BOT_EMOJI)
        await ctx.send_reply(answer)
    finally:
        remove_file_if_exists(image_path)


command = Command(
    name="suporte",
    description="Suporte inteligente do Takeshi usando IA.",
    commands=["suporte", "help", "ajuda"],
    usage="/suporte como instalar o Takeshi?",
    handle=handle,
)

File: src/commands/member/suporte.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import OpenAI from "openai";
import { BOT_EMOJI, OPENAI_API_KEY, PREFIX } from "../../config.js";
import { DangerError, WarningError } from "../../errors/index.js";
import { getRandomName } from "../../utils/index.js";

// Identifiers commonly stuck after ``` by models (```javascript, ```bash, etc).
const CODE_FENCE_LANGUAGES =
  "javascript|js|jsx|typescript|ts|tsx|bash|sh|shell|zsh|json|jsonc|yaml|yml|toml|xml|html|css|scss|go|golang|python|py|ruby|rb|php|java|kotlin|kt|rust|rs|c|cpp|csharp|cs|swift|dart|sql|graphql|md|markdown|diff|dockerfile|docker|powershell|ps1|cmd|bat|ini|env|text|txt|plaintext|vue|svelte|lua|r|perl|scala|nginx|makefile|proto|protobuf|nodejs|node";

/**
 * Only transforms fences like ```javascript / ```bash / ```ts into bare ```.
 * Does not touch language names written in normal prose.
 */
function normalizeWhatsAppCodeBlocks(text) {
  return String(text || "").replace(
    new RegExp(`\`\`\`(?:${CODE_FENCE_LANGUAGES})(?=[\\s\\r\\n]|$)`, "gi"),
    "```",
  );
}

/**
 * Removes unsolicited calls to continue the conversation from prose while
 * preserving code blocks. The system prompt should prevent these, but this is
 * a final guard because the command must answer the current request only.
 */
function removeUnsolicitedFollowUps(text) {
  const continuationPattern =
    /(?:^|\s)(?:se\s+quiser|se\s+preferir|caso\s+queira)[,:]?\s*(?:eu\s+)?(?:posso|te\s+(?:passo|envio|mostro|explico|ajudo)|lhe\s+(?:passo|envio|mostro|explico)|preparo|forneço)\b|^\s*(?:posso\s+(?:também\s+)?(?:te|lhe)\s+(?:passar|enviar|mostrar|explicar)|quer\s+que\s+eu\s+(?:te\s+)?(?:passe|envie|mostre|explique))/i;

  let insideCodeBlock = false;

  return String(text || "")
    .split(/\r?\n/)
    .map((line) => {
      if (line.trimStart().startsWith("```")) {
        insideCodeBlock = !insideCodeBlock;
        return line;
      }

      if (insideCodeBlock) return line;

      const match = line.match(continuationPattern);
      return match ? line.slice(0, match.index).trimEnd() : line;
    })
    .filter((line, index, lines) => line || lines[index - 1] || lines[index + 1])
    .join("\n")
    .trim();
}

export default {
  name: "suporte",
  description: "Suporte inteligente do Takeshi usando IA treinada",
  commands: ["suporte", "help", "ajuda"],
  usage: `${PREFIX}suporte como instalar o Takeshi no Termux?

Você também pode enviar uma imagem com o comando ${PREFIX}suporte

Você também pode escrever o texto e responder a mensagem com o comando ${PREFIX}suporte`,
  /**
   * @param {CommandHandleProps} props
   */
  handle: async ({
    fullArgs,
    args,
    sendReply,
    sendWaitReply,
    sendReact,
    replyText,
    isImage,
    isVideo,
    isAudio,
    downloadImage,
    webMessage,
  }) => {
    if (!OPENAI_API_KEY) {
      throw new WarningError(
        "O suporte inteligente não está disponível no momento. Entre em contato com o administrador do bot!",
      );
    }

    if (isVideo) {
      throw new WarningError(
        "Não consigo interpretar vídeos ainda! Envie uma imagem ou texto!",
      );
    }

    if (isAudio) {
      throw new WarningError(
        "Não consigo interpretar áudios ainda! Envie uma imagem ou texto!",
      );
    }

    const doubleContext = args.length && replyText;
    const text = args.length ? fullArgs : replyText;

    if (!text && !isImage) {
      await sendReact(BOT_EMOJI);

      await sendReply(
        `*Takeshi Suporte*
        
Faça sua pergunta sobre mim que eu te ajudarei!
  
📝 *Exemplos*

- ${PREFIX}suporte bot desliga sozinho
- ${PREFIX}suporte como instalar no Termux?
- ${PREFIX}suporte erro 401 API Spider X
- Envie uma imagem com ${PREFIX}suporte para análise visual`,
      );

      return;
    }

    await sendWaitReply("Analisando sua pergunta...");

    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);

    const finalText = doubleContext
      ? `Contexto anterior: ${replyText}\n\nNova questão: ${text}`
      : text;

    if (finalText) {
      const minLength = 5;
      const maxLength = 2048;

      if (finalText.length < minLength) {
        throw new DangerError(
          `O texto deve ter no mínimo ${minLength} caracteres.`,
        );
      }

      if (finalText.length > maxLength) {
        throw new DangerError(
          `O texto deve ter no máximo ${maxLength} caracteres.`,
        );
      }
    }

    let imagePath = null;

    if (isImage) {
      imagePath = await downloadImage(webMessage, getRandomName());
    }

    const openai = new OpenAI({
      apiKey: OPENAI_API_KEY,
    });

    const messages = [
      {
        role: "system",
        content: `Você é um assistente especializado em suporte técnico do Takeshi Bot.

Responda apenas assuntos relacionados a: tecnologia, programação, desenvolvimento de bots, inteligência artificial, 
machine learning ou assuntos relacionados ao Takeshi Bot.

Responda apenas em português do Brasil.
Seja direto e objetivo nas respostas, salvo se o usuário solicitar explicações mais aprofundadas.

Escreva como alguém que realmente sabe do que está falando e vai direto ao ponto, logo, não escreva demais, apenas o suficiente para ser objetivo. 
Sem frases de abertura do tipo "Claro!", "Ótima pergunta!", "Com certeza!" ou similares. 
Sem encerramento do tipo "Espero ter ajudado!" ou "Qualquer dúvida é só perguntar!". 
Sem travessão (—) para estruturar listas ou ideias. Sem bullet points a menos que seja absolutamente necessário para clareza. 
Evite palavras de enchimento: "importante", "crucial", "fundamental", "robusto", "abrangente". 
Nunca responda de forma genérica quando uma resposta específica é possível. Se a pergunta for vaga, interprete da forma mais útil e responda com substância, não peça esclarecimentos desnecessários. 
Use exemplos concretos quando ajudar a explicar algo. Se tiver uma opinião sobre o assunto, diga, não fique em cima do muro.

Quando receber imagens, analise o conteúdo visual primeiro e interprete-o considerando o contexto técnico do Takeshi Bot.

Se alguém te pedir o link de alguma Host, envie as que você já conhece, 
sem mencionar Pterodactyl, pois os iniciantes não sabem o que é (exceto se perguntarem sobre)!

REGRA DE TAMANHO (obrigatória): a parte em PROSA da resposta deve ter no máximo 3 parágrafos curtos ou 150 palavras, salvo se o usuário pedir explicação aprofundada. Blocos de código NÃO contam nesse limite: inclua sempre o código completo e funcional necessário, mesmo que longo, sem truncar imports, fechamentos ou partes essenciais. Respostas objetivas não precisam de introdução nem de conclusão. Vá direto à solução.

REGRA DE ENCERRAMENTO (obrigatória): entregue somente o que foi solicitado e encerre a resposta ao concluir. Nunca ofereça continuação, ajuda adicional, scripts, métodos, funções, exemplos ou próximos passos que o usuário não pediu. São proibidas frases como "Se quiser, posso...", "Posso te passar...", "Quer que eu...", "Caso queira..." e equivalentes, mesmo como última frase.

REGRA DE CÓDIGO NO WHATSAPP (obrigatória): nunca cole a linguagem no fence.
Errado: \`\`\`javascript  |  \`\`\`bash  |  \`\`\`js  |  \`\`\`ts  |  \`\`\`go
Certo: apenas \`\`\` sozinho na abertura e no fechamento.
Você pode citar a linguagem no texto normal se precisar, mas NUNCA logo após as três crases.

Exemplo correto:
\`\`\`
const x = 1;
\`\`\`

Exemplo errado (não use):
\`\`\`javascript
const x = 1;
\`\`\``,
      },
    ];

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "..", "AGENTS.md"),
        "utf-8",
      ),
    });

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "..", "README.md"),
        "utf-8",
      ),
    });

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "..", "CONTRIBUTING.md"),
        "utf-8",
      ),
    });

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "..", "package.json"),
        "utf-8",
      ),
    });

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "menu.js"),
        "utf-8",
      ),
    });

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "connection.js"),
        "utf-8",
      ),
    });

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "loader.js"),
        "utf-8",
      ),
    });

    messages.push({
      role: "system",
      content: fs.readFileSync(
        path.resolve(__dirname, "..", "..", "@types", "index.d.ts"),
        "utf-8",
      ),
    });

    const userMessage = {
      role: "user",
      content: [],
    };

    if (finalText) {
      userMessage.content.push({
        type: "text",
        text: finalText,
      });
    }

    if (imagePath && fs.existsSync(imagePath)) {
      const buffer = fs.readFileSync(imagePath);
      const base64 = buffer.toString("base64");
      const ext = path.extname(imagePath).toLowerCase();

      let mimeType = "image/jpeg";
      switch (ext) {
        case ".png":
          mimeType = "image/png";
          break;
        case ".jpg":
        case ".jpeg":
          mimeType = "image/jpeg";
          break;
        case ".webp":
          mimeType = "image/webp";
          break;
        case ".gif":
          mimeType = "image/gif";
          break;
      }

      userMessage.content.push({
        type: "image_url",
        image_url: {
          url: `data:${mimeType};base64,${base64}`,
          detail: "low",
        },
      });
    }

    if (!finalText && isImage) {
      userMessage.content.unshift({
        type: "text",
        text: "O que você vê nesta imagem?",
      });
    }

    messages.push(userMessage);

    const response = await openai.chat.completions.create({
      model: "gpt-5.4-mini",
      messages: messages,
      reasoning_effort: "low",
      max_completion_tokens: 2048,
    });

    const answer = removeUnsolicitedFollowUps(
      normalizeWhatsAppCodeBlocks(response.choices[0].message.content.trim()),
    );

    if (!answer) {
      throw new DangerError(
        `Não consegui encontrar uma resposta para sua pergunta. Tente reformular ou ser mais específico!

Não respondo assuntos fora do meu escopo de tecnologia!`,
      );
    }

    await sendReact(BOT_EMOJI);
    await sendReply(answer);

    if (imagePath && fs.existsSync(imagePath)) {
      fs.unlinkSync(imagePath);
    }
  },
};
